### DISCUSSION TITLE
Error: "Attempt to fetch from allocatable variable IMAP_IRR when it is not allocated"

### DISCUSSION THREAD

**Discussion by hellolori** [SSCHAcode/python-sscha#181] [Q&A] [answered]
Dear developers,

I am currently learning how to use EPIq.x to calculate the electron-phonon coupling of doped MoS2, following the tutorial from SSCHA SCHOOL 2023 in the chapter titled "HANDS-ON-SESSION 8: EPIQ". According to the tutorial, I begin with a preliminary step referred to as "dump" using the following input files (dump.in).

```
&control
    dump_gR=.true.,
    prefix='MoS2',
    elphmat_dir='./mat_elem/',
/
&electrons
/
&phonons
 nq1=8,
 nq2=8,
 nq3=1,
/
```

I run the program epiq.x ( version 1.2) using the following command:

```bash
epiq.x -inp dump.out
```

However, the  epiq.x encountered an error and exited. The last few lines of the output are as follows：

```
  Coordinate of Wannier centres and spreads
  wfc #   1   [  -0.00116229  3.23465700  2.11536589 ]  sprd=  1.34007537
  wfc #   2   [   1.57865221  0.44483452  5.19208032 ]  sprd=  1.27174090
  wfc #   3   [   1.58342305  0.90525633  4.54530171 ]  sprd=  1.51677405
  wfc #   4   [   3.04119923  1.85229092  3.44877422 ]  sprd=  1.69789195
  wfc #   5   [  -0.01040035  1.77422646  3.45617143 ]  sprd=  1.74288547
  wfc #   6   [  -1.55362175  4.57155248  3.44622500 ]  sprd=  1.94230737
  wfc #   7   [   0.12418174  1.85329440  3.43774705 ]  sprd=  1.70520003
  wfc #   8   [   1.17252018  1.14910122  5.18003765 ]  sprd=  1.26991476
  wfc #   9   [   1.98577276  1.14867764  5.18587484 ]  sprd=  1.26980273
  wfc #  10   [   1.18074780  1.16967434  1.97254944 ]  sprd=  1.28998609
  wfc #  11   [   3.13819585  1.81165932  3.44660347 ]  sprd=  1.87955423
  wfc #  12   [   1.58103086  0.80995132  1.37619333 ]  sprd=  1.39759096
  wfc #  13   [   1.97921377  1.16956659  1.97430848 ]  sprd=  1.28939906


forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable IMAP_IRR when it is not allocated

Image              PC                Routine            Line        Source
epiq.x             0000000000673F46  Unknown               Unknown  Unknown
epiq.x             000000000060F3CF  read_aux_files_            30  read_aux_files.F90
epiq.x             000000000040C4B1  epiq_                      61  epiq.F90
epiq.x             000000000040C389  MAIN__                     84  header_MPI_epiq.F90
epiq.x             000000000040BEA2  Unknown               Unknown  Unknown
libc-2.17.so       00002B6F4C44C505  __libc_start_main     Unknown  Unknown
epiq.x             000000000040BDA9  Unknown               Unknown  Unknown
```

Can you give some suggestion for me to solve the above issuse?

The version of epiq.x I used is 1.2, which was complied using the following make.sys
```
#=======================
# COMPILER
#=======================
MPF90         =  ~/software/openmpi/4.0.5/bin/mpifort
F90           = ifort

#=======================
# NETLIB LAPACK and BLAS
#=======================
#LIBDIR = /usr/local/lib
#LIBS = -L$(LIBDIR)   -llapack -lblas
LIBS = -L -lmkl_blacs_sgimpt_lp64 -lmkl_intel_lp64  -lmkl_sequential -lmkl_core

#========
# OPTIONS
#========
FCOPTS =  -C   -O2 -assume byterecl -g -traceback
LDOPTS =  -C   -O2 -assume byterecl -g -traceback
```

Thanks a lot!
Feipeng

**Comment by mesonepigreco** [comment]
Hi, thanks for reaching out.
I tag two developers of EPIq:
@giomarini1993 @babbyone 

hope they can help you with this issue or redirect to the EPIq specific documentation.

**Comment by babbyone** [comment]
Hi,
It seems like you manage to compile the code successfully.
However, you show an error that I have never encountered.
I imagine that you run the program as 
`epiq.x -inp dump.in > dump.out`, 
and not as
 `epiq.x -inp dump.out` 
as you wrote, but I had to tell you.

Moreover, you can try to run the program adding explicitly
the "calculation" label, as:
```
&control
    dump_gR=.true.,
    prefix='MoS2',
    elphmat_dir='./mat_elem/',
    calculation='ph_linewidth',
/
&electrons
/
&phonons
 nq1=8,
 nq2=8,
 nq3=1,
/
```

Let us know if this solve your issue,
guglielmo

**Reply by hellolori** [reply]
Dear guglielmo,
Thank you for pointing out the incorrect command I wrote. Actuallty, I run the program using the correct command as you  pointed out:
```bash
epiq.x -inp dump.in > dump.out
```

I have added the  calculation='ph_linewidth' label following you suggestions. However, the aforementioned error still occurs.

Regarding the installation of Epiq, I encountered some difficulties during the compilation process to generate epiq.x. 
First, I followed the instructions and downloaded the installation package.
```bash
git clone --depth 1  --branch main https://gitlab.com/the-epiq-team/epiq.git
```
Afterwards, I compiled epiq.x using the following make.sys:
```
#=======================
# COMPILER
#=======================
MPF90         =  ~/software/openmpi/4.0.5/bin/mpifort
F90           = ifort

#=======================
# NETLIB LAPACK and BLAS
#=======================
#LIBDIR = /usr/local/lib
#LIBS = -L$(LIBDIR)   -llapack -lblas
LIBS = -L -lmkl_blacs_sgimpt_lp64 -lmkl_intel_lp64  -lmkl_sequential -lmkl_core

#========
# OPTIONS
#========
FCOPTS =  -C   -O2 -assume byterecl -g -traceback
LDOPTS =  -C   -O2 -assume byterecl -g -traceback
```
However, I encountered an error message when attempting to compile it directly.
```
ifort  -C   -O2 -assume byterecl -g -traceback  -c io_var.F90
io_var.F90(345): error #6345: Either a PAUSE, STOP or ERROR STOP statement has an invalid argument.   [TRIM]
  STOP TRIM(message)
-------^
compilation aborted for io_var.F90 (code 1)
make[1]: *** [io_var.o] Error 1
make[1]: Leaving directory `/home/hellolori/software/EPIq/1.2/epiq/src'
make: *** [epiq] Error 1
```
I successfully compiled and generated epiq.x by making modifications to io_var.F90, which helped me avoid the above error. The differences between the original file (io_var.ori.F90) and the modified file (io_var.modified.F90) are as follows:
 ```
(base) [hellolori@login src] $ diff io_var.ori.F90 io_var.modified.F90
321a322
>   character(len=100) :: trimmed_message  !!modified
345c346,351
<   STOP TRIM(message)
---
> !!modified
>   trimmed_message = trim(message)
>   write(*,*) "stop, ",trimmed_message
>   stop
> !!modified-end
>   ! STOP TRIM(message)
```

I am not sure if the above approach is related to the current error.

Thanks a lot!
Feipeng

**Reply by hellolori** [reply, marked as answer]
Dear guglielmo,

I think I have worked out the problem. It seems that there is something wrong with the "make.sys" file.
By modifying the value of FGDB from "-C -O2 -assume byterecl -g -traceback" to "-O2 -assume byterecl -g -traceback" in "make.sys", I was able to avoid the aforementioned error and successfully perform the "dump" calculation.

Bests!
Feipeng

**Reply by mesonepigreco** [reply]
Thanks for sharing the solution!
